### PR TITLE
SONARJS-801 Always force coverage to zero when no coverage provided

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 1]
```diff
@@ -342,16 +351,16 @@ public void execute(SensorContext context) {
 
     analyseFiles(context, treeVisitors, fileSystem.inputFiles(mainFilePredicate), progressReport);
 
-    executeCoverageSensors(context, linesOfCode);
+    executeCoverageSensors(context, linesOfCode, sqVersionGreaterOrEqual62);
   }
 
-  private static void executeCoverageSensors(SensorContext context, Map<InputFile, Set<Integer>> linesOfCode) {
+  private static void executeCoverageSensors(SensorContext context, Map<InputFile, Set<Integer>> linesOfCode, boolean greaterThanOrEqual6_2) {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MINOR Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00117)

File: javascript-frontend/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java [hunk 2]
```diff
@@ -159,7 +160,11 @@ private void saveLineMetrics(TreeVisitorContext context) {
 
     FileLinesContext fileLinesContext = fileLinesContextFactory.createFor(this.inputFile);
     for (int line = 1; line <= linesNumber; line++) {
-      fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, linesOfCode.contains(line) ? 1 : 0);
+      int isCodeLine = linesOfCode.contains(line) ? 1 : 0;
+      if (saveExecutableLines) {
+          fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, isCodeLine);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java]: MAJOR Make this line start at column 9. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java [hunk 3]
```diff
@@ -159,7 +160,11 @@ private void saveLineMetrics(TreeVisitorContext context) {
 
     FileLinesContext fileLinesContext = fileLinesContextFactory.createFor(this.inputFile);
     for (int line = 1; line <= linesNumber; line++) {
-      fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, linesOfCode.contains(line) ? 1 : 0);
+      int isCodeLine = linesOfCode.contains(line) ? 1 : 0;
+      if (saveExecutableLines) {
+        fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, isCodeLine);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java]: MAJOR Make this line start at column 9. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 4]
```diff
@@ -323,8 +324,9 @@ public void describe(SensorDescriptor descriptor) {
   public void execute(SensorContext context) {
     List<TreeVisitor> treeVisitors = Lists.newArrayList();
     Map<InputFile, Set<Integer>> linesOfCode = new HashMap<>();
+    boolean sqGreater61 = context.getSonarQubeVersion().isGreaterThanOrEqual(V6_2);
```
Comment (pynicolas) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: @vilchik-elena Suggested name: `isAtLeastSq62`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/LCOVCoverageSensor.java [hunk 5]
```diff
@@ -41,7 +41,7 @@
 abstract class LCOVCoverageSensor {
   private static final Logger LOG = Loggers.get(UTCoverageSensor.class);
 
-
+  private boolean sqGreater61;
```
Comment (pynicolas) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/LCOVCoverageSensor.java]: @vilchik-elena Suggested name: isAtLeastSq62

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java [hunk 6]
```diff
@@ -176,39 +184,76 @@ public void test_unresolved_path() {
...
+
+    utCoverageSensor.execute(context, linesOfCode, RUN_WITH_SQ_6_1);
+    assertThat(logTester.logs()).doesNotContain(message);
+
+    context.setSettings(new Settings().setProperty(FORCE_ZERO_COVERAGE_KEY, "true"));
+    utCoverageSensor.execute(context, linesOfCode, RUN_WITH_SQ_6_1);
+    assertThat(logTester.logs()).doesNotContain(message);
+
+    utCoverageSensor.execute(context, linesOfCode, RUN_WITH_SQ_6_2);
+    assertThat(logTester.logs()).contains(message);
```
Comment (pynicolas) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java]: @vilchik-elena What if we don't set "forceZeroCoverage" on SQ 6.2?

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java [hunk 7]
```diff
@@ -176,39 +184,76 @@ public void test_unresolved_path() {
...
+    assertThat(context.lineHits("moduleKey:file1.js", CoverageType.UNIT, 1)).isNull();
+
+    newSettings.setProperty(JavaScriptPlugin.LCOV_UT_REPORT_PATH, "reports/report_ut.lcov");
+    context.setSettings(newSettings);
+
+    itCoverageSensor.execute(context, linesOfCode, RUN_WITH_SQ_6_2);
+    assertThat(context.lineHits("moduleKey:file1.js", CoverageType.IT, 1)).isNull();
+
+    overallCoverageSensor.execute(context, linesOfCode, RUN_WITH_SQ_6_2);
+    assertThat(context.lineHits("moduleKey:file1.js", CoverageType.OVERALL, 1)).isEqualTo(2);
```
Comment (pynicolas) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java]: @vilchik-elena Why?

Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/lcov/CoverageSensorTest.java]: @pynicolas We test that we save measure on SQ 6.2.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java (hunk 4 above)
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: @pynicolas done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/LCOVCoverageSensor.java [hunk 8]
```diff
@@ -51,10 +51,17 @@ private static FilePredicate mainFilePredicate(FileSystem fileSystem) {
       fileSystem.predicates().hasLanguage(JavaScriptLanguage.KEY));
   }
 
-  public void execute(SensorContext context, Map<InputFile, Set<Integer>> linesOfCode) {
+  public void execute(SensorContext context, Map<InputFile, Set<Integer>> linesOfCode, boolean sqGreater61) {
+    this.sqGreater61 = sqGreater61;
+
+    boolean isForceZeroCoverageActivated = isForceZeroCoverageActivated(context);
+    if (sqGreater61 && isForceZeroCoverageActivated) {
+      LOG.warn("Property 'sonar.javascript.forceZeroCoverage' is removed and its value is not used during analysis");
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/lcov/LCOVCoverageSensor.java]: @pynicolas done

File: javascript-frontend/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java [hunk 9]
```diff
@@ -40,7 +43,7 @@
 import org.sonar.plugins.javascript.api.visitors.SubscriptionVisitor;
 import org.sonar.plugins.javascript.api.visitors.TreeVisitorContext;
 
-public class MetricsVisitor extends SubscriptionVisitor {
+public class MetricsVisitor extends SubscriptionVisitor implements Function<InputFile, Set<Integer>> {
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java]: @vilchik-elena Why? How can we guess what this `Set<Integer>` is about?

**Pull request merged**